File: src-ts/scoring/hash-validator.ts
import { createHash } from "node:crypto";

function isPlainObject(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function sortKeys(value: unknown): unknown {
  if (Array.isArray(value)) {
    return value.map(sortKeys);
  }
  if (isPlainObject(value)) {
    const out: Record<string, unknown> = {};
    for (const key of Object.keys(value).sort()) {
      out[key] = sortKeys(value[key]);
    }
    return out;
  }
  return value;
}

export function canonicalJson(obj: Record<string, unknown>): string {
  return JSON.stringify(sortKeys(obj));
}

export function computeContentHash(receipt: Record<string, unknown>): string {
  const copy = JSON.parse(JSON.stringify(receipt)) as Record<string, unknown>;
  const meta = isPlainObject(copy.meta) ? { ...copy.meta } : undefined;
  if (meta) {
    delete meta.content_hash;
    if (Object.keys(meta).length === 0) {
      delete copy.meta;
    } else {
      copy.meta = meta;
    }
  }

  const canonical = canonicalJson(copy);
  const digest = createHash("sha256").update(canonical, "utf8").digest("hex");
  return `sha256:${digest}`;
}

export function validateContentHash(receipt: Record<string, unknown>): boolean {
  const meta = isPlainObject(receipt.meta) ? receipt.meta : {};
  const claimed = typeof meta.content_hash === "string" ? meta.content_hash : "";
  if (!claimed) return true;
  if (!claimed.startsWith("sha256:")) return false;
  return computeContentHash(receipt) === claimed;
}

File: src/scoring/hash_validator.py
"""
Content hash computation and validation for ship-receipts.

Spec: docs/specs/ship-receipts-scoring-model-v1.md §3
"""

from __future__ import annotations

import hashlib
import json


def canonical_json(obj: dict) -> str:
    """Produce canonical JSON: sorted keys, compact separators, UTF-8."""
    return json.dumps(obj, sort_keys=True, separators=(",", ":"), ensure_ascii=False)


def compute_content_hash(receipt: dict) -> str:
    """
    Compute SHA-256 of the canonical receipt JSON, excluding meta.content_hash.

    Returns "sha256:<hex>".
    """
    if not isinstance(receipt, dict):
        raise TypeError("receipt must be a dict")

    receipt_copy = json.loads(json.dumps(receipt))
    meta = receipt_copy.get("meta")
    if isinstance(meta, dict):
        meta.pop("content_hash", None)
        if not meta:
            receipt_copy.pop("meta", None)

    canonical = canonical_json(receipt_copy)
    digest = hashlib.sha256(canonical.encode("utf-8")).hexdigest()
    return f"sha256:{digest}"


def validate_content_hash(receipt: dict) -> bool:
    """
    Check that meta.content_hash matches the computed hash.

    Returns True if:
    - content_hash is present and matches
    Returns False if:
    - content_hash is present but does NOT match
    Returns True if:
    - content_hash is absent (nothing to validate)
    """
    if not isinstance(receipt, dict):
        return False

    meta = receipt.get("meta")
    if meta is None:
        return True  # No hash to validate
    if not isinstance(meta, dict):
        return False

    claimed = meta.get("content_hash", "")
    if not claimed:
        return True  # No hash to validate

    if not isinstance(claimed, str):
        return False
    if not claimed.startswith("sha256:"):
        return False

    computed = compute_content_hash(receipt)
    return computed == claimed
